### PR TITLE
Session activity timeout

### PR DESCRIPTION
**Pull request by calherries**
[Product doc](https://www.notion.so/metabase/Offer-more-control-over-session-timeouts-bc8e84e49de44de0950385fc839b8df3)
Fixes https://github.com/metabase/metabase-enterprise/issues/326

## Changes

This PR adds the new session timeout setting. It defines a timespan after which users should be logged out. This means all open tabs will be redirected to the login screen automatically on timeout.

## Technical details

Session timeout functionality is added as middleware on the backend and the frontend. 

The backend middleware adds a new cookie `metabase.TIMEOUT` on login, which is a non http-only clone of the `metabase.SESSION` cookie. On every request following login, the `expires` attribute of these two cookies will be updated according to the session-timeout setting if the cookies haven't expired by the time of the request.

On the frontend, every open tab will independently and repeatedly check the presence of the `metabase.TIMEOUT` cookie, which will log the user out when it expires.

## How to verify
- Run Metabase EE
- Open a couple of tabs with random questions and dashboards and keep them in background
- In a new tab open Admin Settings -> [Authentication](http://localhost:3000/admin/settings/authentication)
- Set the session timeout to one minute
- Wait for one minutes without doing any new requests to the backend
- Ensure all the tabs show the login screen after the timeout
- Login back in any of the tabs
- Ensure all tabs are logged in and show the previous url you had before the logout

TODO: 
- [x] Add setting for timeout duration in minutes on BE
- [x] Set this setting from the FE
- [x] Log out from all active tabs on the FE after the timeout

### CODE REVIEW

**Comment by calherries** [comment]
@alxnddr Thanks for doing the FE work for this, I've merged your PR https://github.com/metabase/metabase/pull/23591# into this one.

The backend should be ready for the final touches on the FE. It's successfully logging out the FE by clearing the metabase.SESSION cookie, but the polling behaviour from the frontend isn't working yet. The FE will need to poll the backend to trigger the logout. I assume you haven't tried implementing that part yet.

The only thing to work out on the BE are the caching details described [here](https://www.notion.so/metabase/Offer-more-control-over-session-timeouts-bc8e84e49de44de0950385fc839b8df3#a94d2aeaaedc41f5a25141138a7e22a6).

The big thing I've changed on the BE is the setting is now called `setting-timeout-minutes`, which expects to be set as an integer in minutes. If the setting is nil (which it is by default), then the timeout is not active. You'll have to modify `SessionTimeoutSetting` to set this.

**Comment by calherries** [comment]
I've just handcoded the timeout to 10 seconds for testing the FE logging out properly. Feel free to revert.

This gets at another question. I'm scratching my head on how to do e2e testing for this? 
@dpsutton Got any advice on how to test this properly?

**Comment by alxnddr** [comment]
@calherries the cookie polling behavior is already implemented here https://github.com/metabase/metabase/pull/23349/files#diff-9f0a2e6953dbe29dad402528280e9373b7e7093b89b3877f0ee9d47303a491e9 

FE does not need to pool anything in this case. On every request the BE sets a non-httponly cookie with a new expiration date to prolong the session. When a user doing any activity in Metabase the session will be extended. Once a user stops interacting with a tab, there will be no requests to BE and after the expiration date set by the last request to BE the cookie expires and the FE logs out all tabs

**Comment by alxnddr** [comment]
`metabase.SESSION` cannot be used because it is a httponly cookie. We should not make it non-httponly for security reasons. So we need to set another non-httponly cookie. The value can be anything (like `alive` text). But the expiration date of the cookie should be `currentTime + sessionTimeout`

**Comment by calherries** [comment]
Right, this makes sense. I was struggling to understand how the `session-listener` code worked before, but now it's clear. Cheers. I'll update the BE to get this working.

**Comment by calherries** [comment]
@alxnddr I changed the cookie name to `metabase.TIMEOUT`, and updated the backend to reset the cookie's expiry date on every API request. The only thing left is I'll need you to update the `SessionTimeoutSetting` UI to respond to the backend with `session-timeout-seconds`, defined [here](https://github.com/metabase/metabase/blob/session-timeout/src/metabase/server/middleware/session.clj#L321). This is where you can change the minimum timeout from 60 seconds to something shorter for testing.

I'm still not sure how to test this end to end, however. Currently I've only written a couple of unit tests for the backend.

File: src/metabase/server/middleware/session.clj [hunk 1]
```diff
@@ -305,3 +307,89 @@
...
+
+(defn session-timeout->seconds
+  "Convert a session timeout setting to seconds."
+  [{:keys [unit amount]}]
+  (when amount
+    (-> (case unit
+          "seconds" amount
+          "minutes" (* amount 60)
+          "hours"  (* amount 3600))
+        (max 60)))) ; Ensure a minimum of 60 seconds so a user can't lock themselves out
```
Comment (dpsutton) [inline src/metabase/server/middleware/session.clj:330]: The use of `(max 60 (* amount 60))` seems pretty off here. Did you mean to use `min` here instead?

File: src/metabase/server/middleware/session.clj [hunk 2]
```diff
@@ -305,3 +307,89 @@
...
+                            {:same-site :none
+                             :secure    true}))]
+      (-> response
+          (wrap-body-if-needed)
+          (response/set-cookie metabase-session-timeout-cookie "alive" cookie-options)))
+
+    ;; Clear the session cookie if the timeout cookie expired
+    (and (not (get-in request [:cookies metabase-session-timeout-cookie :value]))
+         (get-in request [:cookies metabase-session-cookie :value]))
+    (clear-session-cookie response)
```
Comment (dpsutton) [inline src/metabase/server/middleware/session.clj:377]: If I'm reading this right, if a request lacks a timeout cookie it will log the user out. This seems like a rough way to implement this. For example, anyone who uses Metabase's API will have quite the confusing time figuring out why everything is failing when they upgrade. I'm also a fan of using the API `http POST localhost:3000/... -Cookie $COOKIE`. Now it seems if i don't continually update that cookie I'll get logged out. It seems this will be true regardless of whether I use this new feature or not.

**Comment by dpsutton** [comment]
I think that conceptually I had a bit different of an idea.

At the base, this treats information from a client as authoritative and we take action on that. I'm fundamentally wary of things like that.

Some things that seem strange to me in this implementation, even on its own terms:
- this sets a cookie who's value is "alive" when the feature is not needed. Why even bother with this cookie in that case?
- this operates on a well-formed response. This means that if it intends to log someone out, it does it after the user's request has succeeded. I think this is a bit fatal for these purposes. We should really reject the request if it has timed out. Instead, we are log them out after their action has happened.
- This changeset uses a dummy value "alive" and making the cookie expire. Why did you not choose to make the actual session cookie expire?
- I wish the logic was co-located with the session check. It seems like a more elegant solution would check the two cookies in tandem rather than separately like this. Perhaps this would allow for rejecting the request at the beginning of the request lifecycle rather than after the request has already finished.

In my head I had envisioned a solution where we gave the client information about when they should log out but the backend would ensure that the session was deleted regardless of action taken by the client. So I'm surprised to see the backend no longer keeps track of the last activity for a session. The original idea was to basically use a regular header that would return when the client should consider themselves logged out (and the backend would ensure that they were logged out). We use a cookie so that all tabs associated with that session would see the value.

We also need to consider how this affects people who use the Metabase API. Since we don't keep any information on the server any consumers are required to continually keep the latest cookies up to date. This is a big change from currently where they must save only the initial cookie or even use a passed-in session id. I'm also not clear on what can be edited and what cookies need to be sent. Since we don't ever check the timestamp and just rely on it having a value, could anyone construct the cookie with the value "alive" and no expiry and things just work?

File: src/metabase/server/middleware/session.clj (hunk 2 above)
Comment (alxnddr) [reply src/metabase/server/middleware/session.clj:377]: @calherries @dpsutton BE should not check `metabase.TIMEOUT` to decide if it needs to remove the session cookie — everyone can manually edit `metabase.TIMEOUT` and make it not expire after the `Expires` value came from server. Server should basically update `Expires`  value of `metabase.SESSION`, `metabase.TIMEOUT` on every request by `now + sessionTimeout`

Comment (alxnddr) [reply src/metabase/server/middleware/session.clj:377]: Overall, BE should keep the session expiration date and should not rely on any FE data. Otherwise, by using devtools I could extend my session indefinitely

File: src/metabase/server/middleware/session.clj [hunk 3]
```diff
@@ -305,3 +307,89 @@
...
+          "minutes" (* amount 60)
+          "hours"  (* amount 3600))
+        (max 60)))) ; Ensure a minimum of 60 seconds so a user can't lock themselves out
+
+(defn response-with-session-timeout-cookie
+  "Adds a cookie to the response that expires after `session-timeout-seconds` seconds."
+  [request request-time timeout-seconds response]
+  (cond
+    ;; If there is no session cookie because the user is logged out, clear the timeout cookie
+    (and (get-in request [:cookies metabase-session-timeout-cookie :value])
```
Comment (alxnddr) [inline src/metabase/server/middleware/session.clj:337]: It should set the cookie only if the BE session is alive and when it gets removed it should remove the cookie

**Comment by calherries** [comment]
@dpsutton I had not considered the Metabase API. The current implementation assumes a timeout cookie on every authenticated request. But I can avoid that requirement by setting the session cookie to expire at the same time as the timeout cookie, so the timeout cookie is only used for the FE's logout timer. Once that's changed, API users should be able to just use the session-id in the header like before e.g. `"X-Metabase-Session: 38f4939c-ad7f-4cbe-ae54-30946daf8593"` and everything will still work as described in the docs.

On your other considersations, I figured it was unnecessary to maintain the timeout on the backend to achieve the desired functionality. My blind spot is whether this would meet our customer's compliance needs e.g. with [HIPAA](http://whatishipaa.org/). Going through your points:
1. this sets a cookie who's value is "alive" when the feature is not needed. Why even bother with this cookie in that case?
> We need the timeout cookie for the sole purpose of allowing the FE to check whether the session has timed out to log all idle tabs out. The value doesn't matter as the `expires` attribute is sufficient. The session cookie has the `http-only` attribute in its header, which means it cannot be accessed from JS. So we need another cookie.
2. this operates on a well-formed response. This means that if it intends to log someone out, it does it after the user's request has succeeded. I think this is a bit fatal for these purposes. We should really reject the request if it has timed out. Instead, we are log them out after their action has happened.
> You're right, I had not considered this. I can fix this by setting the session cookie to expire at the same time as the timeout cookie.
3. This changeset uses a dummy value "alive" and making the cookie expire. Why did you not choose to make the actual session cookie expire?
> See answer to 1, why we can't just make the session cookie expire. The plan is to make both session and timeout cookies expire at the same time (not implemented yet).
4. I wish the logic was co-located with the session check. It seems like a more elegant solution would check the two cookies in tandem rather than separately like this. Perhaps this would allow for rejecting the request at the beginning of the request lifecycle rather than after the request has already finished.
> Good call. My initial idea was to separate them so as to reason about them independently, but it is clear they are interdependent so I will co-locate their logic.

File: src/metabase/server/middleware/session.clj (hunk 1 above)
Comment (calherries) [reply src/metabase/server/middleware/session.clj:330]: My idea here was to prevent disaster if the duration in seconds was less than a minute, e.g. 0. It's kind of lying to the user doing it this way though, so maybe it's best remove this and add validation on the FE?

File: src/metabase/server/middleware/session.clj (hunk 3 above)
Comment (calherries) [reply src/metabase/server/middleware/session.clj:337]: > when it gets removed it
Can you be more specific on `it` here? Do you mean when the session cookie is removed from logging out? I intend to refactor this so the session cookie gets removed in the same function as the timeout cookie on logout.

Comment (alxnddr) [reply src/metabase/server/middleware/session.clj:337]: I mean the back end should set cookie after verifying that a BE session exists and only in this case. It should not check the existence of other cookies. Maybe I misunderstood the code here because I don't know Clojure at all

File: src/metabase/server/middleware/session.clj (hunk 2 above)
Comment (calherries) [reply src/metabase/server/middleware/session.clj:377]: Loud and clear, I implemented this incorrectly. Just to confirm my understanding: metabase.TIMEOUT can be tampered with from the FE, but metabase.SESSION can't. So metabase.TIMEOUT should only be used to implement the "log out immediately when the timeout expires" feature, which needs to be triggered from the FE (no way around that as I'm aware), but metabase.SESSION will prevent users from making additional API requests from the browser after the timeout expires.

Comment (calherries) [reply src/metabase/server/middleware/session.clj:377]: I just answered my own question. You can change the expiry date and value of all cookies from devtools. I'm now thinking that we need to encrypt the expiry dates of the cookies in the cookie's value so a user can't set their own. What do you think?

Comment (calherries) [reply src/metabase/server/middleware/session.clj:377]: Conor confirmed protecting users from extending sessions with devtools is not a must-have in Notion [here](https://www.notion.so/Offer-more-control-over-session-timeouts-bc8e84e49de44de0950385fc839b8df3?d=91ef8238970b432fb5110ba3a5e19be8#19f1a19389a345e2ab5e388db1078975).

**Comment by calherries** [comment]
The metabase.SESSION and metabase.TIMEOUT cookies now get their `expires` attribute set and updated together, and the cookies themselves are set and cleared together. They only differ in that metabase.SESSION is http-only so can't be tampered with from JS, while metabase.TIMEOUT is not.

**Comment by alxnddr** [comment]
I'm unable to login after I logout on the latest commit, could you please have look?

**Comment by calherries** [comment]
@alxnddr Should work now

File: src/metabase/server/middleware/session.clj (hunk 1 above)
Comment (dpsutton) [reply src/metabase/server/middleware/session.clj:352]: oh you are right. i had it backwards

File: enterprise/frontend/src/metabase-enterprise/auth/middleware/session-middleware.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      intervalId = setInterval(() => {
+        const isLoggedIn = getIsLoggedIn();
+
+        if (isLoggedIn !== wasLoggedIn) {
+          wasLoggedIn = isLoggedIn;
+
+          if (isLoggedIn) {
+            const params = new URLSearchParams(window.location.search);
+            const redirectUrl = params.get("redirect");
+            window.location.replace(redirectUrl ?? "/");
```
[the remark this answers is not in the document]
Comment (alxnddr) [reply enterprise/frontend/src/metabase-enterprise/auth/middleware/session-middleware.js]: I've fixed it by checking the origin

**Review by gusaiani — APPROVED** [review]
Looks good, only one mild thing.

This happened to me:

1. Started EE
2. Opened a few tabs
3. Went to http://localhost:3000/admin/settings/authentication
4. Toggled Session Timeout to enabled, no other change

Immediately, the tab I was in, and all the others, got redirected to http://localhost:3000/

This did not happen later, either by toggling off then on, or by restarting the server and redoing everything.

May be worth trying to reproduce somewhere else.

**Comment by calherries** [comment]
@gusaiani I was able to reproduce your issue with your steps but only if I:

1. Make sure the metabase.TIMEOUT cookies are cleared
2. Start a the dev server on the `master` branch
3. Log in
4. Checkout to the `session-timeout` branch
5. Follow your steps 1-4

The problem comes from the client thinking the user is logged out because of the absence of the `metabase.TIMEOUT` cookie when you switch from one branch to another without logging in again, even if the user is logged in. When you toggled the session timeout to enabled, the `metabase.TIMEOUT` cookie was set for the first time, which is supposed to happen on login. The redux middleware mistakenly thought you just logged in on another tab, and redirected you to the last visited URL, which by default is the homepage.

So your issue would be limited to users that switch versions without logging out, and then toggle the timeout setting.

Anyway, the commit above fixes this. The FE will only redirect the user when the `metabase.TIMEOUT` cookie first appears if there's a `redirect` query parameter in the URL.

**Comment by dpsutton** [comment]
I've run into what I think might be a serious bug, but perhaps it won't happen much in real life usage.

I set the session timeout to 1 minute. Did not do any more activity and was successfully logged out. I then logged back in, turned _off_ the session timeout and then continued to use the app. But I was logged out as I was using the application for one minute with no period of inactivity.

This is because we set on the client a ticker of when to log off and we update that expiration time on each request. When you turn off the setting, nothing tells the client to ignore its ticker, and now nothing keeps resetting the timer.

This seems like an infrequently hit edge but wanted to raise it here. Otherwise seems very nice for the v1 of this feature.

**Comment by calherries** [comment]
@dpsutton Thanks for raising that, the last commit fixes that edge case and adjusts the tests.

**Review by dpsutton — APPROVED** [review]
